### PR TITLE
Refactor uninstall-script generation

### PR DESCRIPTION
**Pull request by vector-of-bool**
This aside change refactors and changes a few minor issues with the uninstall script:

1. The script did not handle a late-defined install-prefix, and entered an infinite loop in that case
2. The script handled `DESTDIR` strangely (?)
3. There were two different generators: for Windows and for Unix
4. Using .bat scripting *at all* is just fraught with peril and should be avoided, if at all possible

This change also removes the top-level `generate_uninstall/` directory, as well as the two generator scripts, instead replacing it with a single `GenerateUninstalle.cmaker` module (which *should* also be portable between projects).

### CODE REVIEW

File: build/cmake/GeneratePkgConfig.cmake
Comment (rcsanchez97) [inline build/cmake/GeneratePkgConfig.cmake:1]: I am assuming that the DEBUG messages were meant to be left in. I don't have a problem with them (I rather like the extra output), but I wanted to mention it in the event you intended to remove them and forgot.

File: build/cmake/GenerateUninstaller.cmake [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        FILES "${_uninstaller_script}"
+        DESTINATION "${UNINSTALL_PROG_DIR}"
+        PERMISSIONS
+            OWNER_READ OWNER_WRITE OWNER_EXECUTE
+            GROUP_READ GROUP_EXECUTE
+            WORLD_READ WORLD_EXECUTE
+        )
+
+    # If applicable, generate an "uninstall" target to run the uninstaller:
+    if(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR OR PROJECT_IS_TOP_LEVEL)
```
Comment (rcsanchez97) [inline build/cmake/GenerateUninstaller.cmake:46]: Is this how a collision is prevented between the uninstall generated in this project and, for instance, the uninstall generated in the C++ driver when it includes the C driver directly?

File: build/cmake/GenerateUninstaller.cmake [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    file(RELATIVE_PATH relpath "${install_prefix}" "${installed}")
+    # Add a removal:
+    add_rmfile("${relpath}")
+    # Climb the path and collect directories:
+    while("1")
+        get_filename_component(installed "${installed}" DIRECTORY)
+        file(TO_NATIVE_PATH "${installed}" installed)
+        get_filename_component(parent "${installed}" DIRECTORY)
+        file(TO_NATIVE_PATH "${parent}" parent)
+        # Don't account for the prefix or directly children of the prefix:
```
Comment (rcsanchez97) [inline build/cmake/GenerateUninstaller.cmake:253]:
```suggestion
        # Don't account for the prefix or direct children of the prefix:
```

File: build/cmake/GenerateUninstaller.cmake [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            break()
+        endif()
+        # Keep track of this directory for later:
+        list(APPEND dirs_to_remove "${installed}")
+    endwhile()
+endforeach()
+
+# Now generate commands to remove (empty) directories:
+list(REMOVE_DUPLICATES dirs_to_remove)
+# Order them in descending so that we remove subdirectories before their parents:
```
Comment (rcsanchez97) [inline build/cmake/GenerateUninstaller.cmake:264]:
```suggestion
# Order them by depth so that we remove subdirectories before their parents:
```

**Review by rcsanchez97 — APPROVED** [review]
I'm a little sad that you didn't like my generator script written in batch :disappointed: 

Seriously, though, this is much nicer. Aside from a few minor tweaks/questions, it looks good to me.

**Review by kkloberdanz — APPROVED** [review]

File: build/cmake/GeneratePkgConfig.cmake
Comment (vector-of-bool) [reply build/cmake/GeneratePkgConfig.cmake:1]: Yes, they were useful while debugging and don't get in the way, so I just left them there.

File: build/cmake/GenerateUninstaller.cmake (hunk 1 above)
Comment (vector-of-bool) [reply build/cmake/GenerateUninstaller.cmake:46]: Yes. This is mostly the same as what was done previously, but with the addition of the newer CMake feature `PROJECT_IS_TOP_LEVEL`, which was added to address this exact use case (on older CMake, that variable check will be a no-op).

**Pull request merged**